### PR TITLE
Fix ImGui keybindings screen not allowing the user to scroll the list upon opening

### LINKED ISSUE #72066: ImGui Keybindings window. Inactive search bar
### Describe the bug

The keybindings window (settings / keybindings). The search string does not become active.

To be more precise, it sometimes becomes inactive (does not allow you to enter text). The behavior is as follows. When opening a window, the cursor blinks once and disappears. When pressing keys nothing is typed.

The interface must be enabled. ```Use imGui UI: true```

### Attach save file

N/A

### Steps to reproduce

1. Default settings
2. Enable imGui (settings / options / interface / Use imGui UI: true)
3. Open the keybindings menu
4. Try typing. If you succeed, go to the next step.
5. Close the game and open it again.
6. Go to step 3

Alternatively 1.
1. Open the keybindings menu
2. Close the keybindings menu
3. Open another menu (safemode, options or create a world)
4. Close this menu
5. Go to step 1

Alternatively 2.
After opening the keybindings menu, press enter. Try typing in the text

### Expected behavior

The search bar should always be active.

### Screenshots

_No response_

### Versions and configuration

Game version: cdda-windows-tiles-x64-2024-02-28-2115

### Additional context

If you disable imGui, everything works fine.

### LINKED ISSUE #72598: [imgui] Can't navigate list in keybindings UI with the keyboard
### Describe the bug

There is no way to navigate the list of keybindings without a mouse

### Attach save file

N/A

### Steps to reproduce

1. Open keybindings UI
2. Press Up, Down, Page Up, Page Down
3. Nothing happens

### Expected behavior

List navigates up and down like in the old UI

### Screenshots

N/A

### Versions and configuration

- OS: Linux
    - OS Version: LSB Version: n/a; Distributor ID: Arch; Description: Arch Linux; Release: rolling; Codename: n/a; 
- Game Version: cdda-experimental-2024-03-24-0121 cf9267f [64-bit]
- Graphics Version: Tiles
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Portal Storms Ignore NPCs [personal_portal_storms],
    Slowdown Fungal Growth [no_fungal_growth]
]

### Additional context

N/A

### PR DESCRIPTION
**Pull request by katemonster33**
#### Summary
Bugfixes "Fixes ImGui keybindings UI not allowing the user to scroll"

#### Purpose of change
Fixes: #72598 
Fixes: #72066

#### Describe the solution
Implemented actions in the keybindings screen to switch to "filtering" mode and back. This breaks from the original behavior of the screen because ImGui does not behave well with having the user use the arrow keys to scroll one part of the screen while also having a text field accepting text input on another part of the screen. Only one control can truly have "focus" at a time and switching focus between elements in order to have user input apply in both places like the old screen, simply is not possible.

#### Describe alternatives you've considered
Modifying ImGui itself to allow programmatic focusing of elements be truly "seamless" so when a user presses an arrow key, have the scrollable element focus, or when they press an alphanumeric key, focus the text field. This is currently possible but at the cost of ImGui "eating" the input that the user makes that causes the focus switch. There is simply no way around it.

#### Additional context
Turns out that modifying the text passed to the InputText field programmatically is a very annoying operation, and possibly requires a new version of ImGui in order to work properly.

### CODE REVIEW

File: src/cata_imgui.cpp [hunk 1]
```diff
@@ -461,6 +462,13 @@ class cataimgui::window_impl
         }
 };
 
+class cataimgui::filter_box_impl
+{
+    public:
+        char text[255]; // NOLINT(modernize-avoid-c-arrays)
```
Comment (akrieger) [inline src/cata_imgui.cpp:468]: I mean, c'mon.
```suggestion
        std::array<char, 255> text;
```

File: src/cata_imgui.cpp [hunk 2]
```diff
@@ -599,3 +607,60 @@ cataimgui::bounds cataimgui::window::get_bounds()
...
+        ImGui::SameLine();
+    } else {
+        action_button( "QUIT", ctxt.get_button_text( "QUIT", _( "Cancel" ) ) );
+        ImGui::SameLine();
+        action_button( "TEXT.CONFIRM", ctxt.get_button_text( "TEXT.CONFIRM", _( "OK" ) ) );
+        ImGui::SameLine();
+    }
+    ImGui::BeginDisabled( filtering_active );
+    ImGui::InputText( "##FILTERBOX", filter_impl->text,
+                      std::extent_v < decltype( filter_impl->text ) > );
```
Comment (akrieger) [inline src/cata_imgui.cpp:632]:
```suggestion
                      filter_impl->text.size() );
```

File: src/cata_imgui.cpp [hunk 3]
```diff
@@ -599,3 +607,60 @@ cataimgui::bounds cataimgui::window::get_bounds()
...
+    ImGui::EndDisabled();
+    if( !filter_impl->id ) {
+        filter_impl->id = GImGui->LastItemData.ID;
+    }
+}
+
+std::string cataimgui::window::get_filter()
+{
+    if( filter_impl ) {
+        return std::string( filter_impl->text );
```
Comment (akrieger) [inline src/cata_imgui.cpp:642]:
```suggestion
        return std::string( filter_impl->text.data() );
```

File: src/cata_imgui.cpp [hunk 4]
```diff
@@ -599,3 +607,60 @@ cataimgui::bounds cataimgui::window::get_bounds()
...
+}
+
+//void cataimgui::window::set_filter( const std::string &filter )
+//{
+//    // doesnt currently work, relies on API only available in newer ImGUi, because I can't have nice things
+//    //ImGuiInputTextState* input_state = ImGui::GetInputTextState( p_impl->id );
+//    //if( input_state ) {
+//    //    input_state->ReloadUserBufAndSelectAll();
+//    //}
+//}
```
Comment (akrieger) [inline src/cata_imgui.cpp:666]: Can we just, not commit this and the other commented out functions?

**Review by akrieger — APPROVED** [review]
Please delete the dead code (especially the newly added dead code).

Also tidy is still sad.

File: src/cata_imgui.cpp (hunk 1 above)
Comment (katemonster33) [reply src/cata_imgui.cpp:468]: the c-style array here came from input_context.cpp where I have vivid memories of trying to do this in the original ImGui PR and it not working and having to fall back on this. But I will give it another try

File: src/cata_imgui.cpp (hunk 4 above)
Comment (katemonster33) [reply src/cata_imgui.cpp]: done

File: src/cata_imgui.cpp (hunk 3 above)
Comment (katemonster33) [reply src/cata_imgui.cpp]: done

File: src/cata_imgui.cpp (hunk 2 above)
Comment (katemonster33) [reply src/cata_imgui.cpp]: done

File: src/cata_imgui.cpp (hunk 1 above)
Comment (katemonster33) [reply src/cata_imgui.cpp]: done, not sure why we nuked that before

**Pull request merged**